power_grid_model_c/power_grid_model/include/power_grid_model/common/logging_impl.hpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "logging.hpp"
+
+#include <string_view>
+#include <vector>
+
+namespace power_grid_model::common::logging {
+class NoLogger : public Logger {
+    void log(LogEvent /*tag*/) override {}
+    void log(LogEvent /*tag*/, std::string_view /*message*/) override {}
+    void log(LogEvent /*tag*/, double /*value*/) override {}
+    void log(LogEvent /*tag*/, Idx /*value*/) override {}
+    std::unique_ptr<Logger> clone() const override { return std::make_unique<NoLogger>(); }
+};
+
+class LogDispatcher final : public LogDispatch {
+  public:
+    void log(LogEvent tag) override { log_impl(tag); }
+    void log(LogEvent tag, std::string_view message) override { log_impl(tag, message); }
+    void log(LogEvent tag, double value) override { log_impl(tag, value); }
+    void log(LogEvent tag, Idx value) override { log_impl(tag, value); }
+    void registrar(Logger* logger) override {
+        if (logger != nullptr && std::ranges::find(loggers_, logger) == loggers_.end()) {
+            loggers_.push_back(logger);
+        }
+    }
+    void deregistrar(Logger* logger) override {
+        if (logger != nullptr) {
+            std::erase(loggers_, logger);
+        }
+    }
+    std::unique_ptr<Logger> clone() const override { return std::make_unique<LogDispatcher>(); }
+
+    LogDispatcher() = default;
+    LogDispatcher(const LogDispatcher&) = delete;
+    LogDispatcher& operator=(const LogDispatcher&) = delete;
+    LogDispatcher(LogDispatcher&&) = default;
+    LogDispatcher& operator=(LogDispatcher&&) = default;
+    ~LogDispatcher() override = default;
+
+  private:
+    std::vector<Logger*> loggers_;
```

Comment:
If multithreading related logic for log was to be moved to logger, it can be:
`std::vector<std::vector<Logger*>> loggers_;`
And there would be a separate aggregation function.

Reply:
indeed. for now, the log dispatcher was removed, but let's keep this in mind indeed